azurerm/resource_arm_analysis_services_server.go

```diff
@@ -0,0 +1,403 @@
+	}
+
+	return warnings, errors
+}
+
+func validateQuerypoolConnectionMode() schema.SchemaValidateFunc {
+	connectionModes := make([]string, len(analysisservices.PossibleConnectionModeValues()))
+	for i, v := range analysisservices.PossibleConnectionModeValues() {
+		connectionModes[i] = string(v)
+	}
+
+	return validation.StringInSlice(connectionModes, true)
+}
+
+func expandAnalysisServicesServerProperties(d *schema.ResourceData) *analysisservices.ServerProperties {
+	adminUsers := expandAnalysisServicesServerAdminUsers(d)
+
+	serverProperties := analysisservices.ServerProperties{AsAdministrators: adminUsers}
+
+	serverProperties.IPV4FirewallSettings = expandAnalysisServicesServerFirewallSettings(d)
+
+	if querypoolConnectionMode, ok := d.GetOk("querypool_connection_mode"); ok {
+		serverProperties.QuerypoolConnectionMode = analysisservices.ConnectionMode(querypoolConnectionMode.(string))
+	}
+
+	return &serverProperties
+}
+
+func expandAnalysisServicesServerMutableProperties(d *schema.ResourceData) *analysisservices.ServerMutableProperties {
+	adminUsers := expandAnalysisServicesServerAdminUsers(d)
+
+	serverProperties := analysisservices.ServerMutableProperties{AsAdministrators: adminUsers}
+
+	serverProperties.IPV4FirewallSettings = expandAnalysisServicesServerFirewallSettings(d)
+
+	serverProperties.QuerypoolConnectionMode = analysisservices.ConnectionMode(d.Get("querypool_connection_mode").(string))
+
+	return &serverProperties
+}
+
+func expandAnalysisServicesServerAdminUsers(d *schema.ResourceData) *analysisservices.ServerAdministrators {
+	adminUsers := d.Get("admin_users").(*schema.Set)
+	members := make([]string, 0)
+
+	for _, admin := range adminUsers.List() {
+		if adm, ok := admin.(string); ok {
+			members = append(members, adm)
+		}
+	}
+
+	return &analysisservices.ServerAdministrators{Members: &members}
+}
+
+func expandAnalysisServicesServerFirewallSettings(d *schema.ResourceData) *analysisservices.IPv4FirewallSettings {
+	firewallSettings := analysisservices.IPv4FirewallSettings{
+		EnablePowerBIService: utils.Bool(d.Get("enable_power_bi_service").(bool)),
+	}
+
+	firewallRules := d.Get("ipv4_firewall_rule").([]interface{})
+
+	fwRules := make([]analysisservices.IPv4FirewallRule, len(firewallRules))
+	for i, v := range firewallRules {
+		fwRule := v.(map[string]interface{})
+		fwRules[i] = analysisservices.IPv4FirewallRule{
+			FirewallRuleName: utils.String(fwRule["name"].(string)),
+			RangeStart:       utils.String(fwRule["range_start"].(string)),
+			RangeEnd:         utils.String(fwRule["range_end"].(string)),
+		}
+	}
+	firewallSettings.FirewallRules = &fwRules
+
+	return &firewallSettings
+}
+
+func flattenAnalysisServicesServerFirewallSettings(serverProperties *analysisservices.ServerProperties) (enablePowerBi *bool, fwRules []interface{}) {
+	if serverProperties.IPV4FirewallSettings == nil {
```

Comment:
there's a potential crash here if `serverProperties` is nil - so could we make this:

```suggestion
	if serverProperties == nil || serverProperties.IPV4FirewallSettings == nil {
```